File: packages/hq-tools/src/tools/vault.ts
import * as fs from "fs";
import * as path from "path";
import matter from "gray-matter";
import { Type } from "@sinclair/typebox";
import type { HQTool, HQContext } from "../registry.js";

const EMBEDDING_MODEL = "openai/text-embedding-3-small";

/** Resolve a vault-relative path and verify it stays inside the vault root. */
function resolveVaultPath(vaultPath: string, relPath: string): string {
  const resolvedVault = path.resolve(vaultPath);
  const fullPath = path.resolve(vaultPath, relPath);
  if (!fullPath.startsWith(resolvedVault + path.sep) && fullPath !== resolvedVault) {
    throw new Error(`Path traversal outside vault is not allowed: ${relPath}`);
  }
  return fullPath;
}

/**
 * Lazy SearchClient pattern — uses dynamic import so hq-tools doesn't need
 * a hard dependency on @repo/vault-client. Host can also inject via ctx.
 */
const clientCache = new Map<string, any>();
async function getSearchClient(ctx: HQContext): Promise<any> {
  if (ctx.searchClient) return ctx.searchClient;
  const cached = clientCache.get(ctx.vaultPath);
  if (cached) return cached;
  // Dynamic import avoids a build-time dependency on vault-client
  const { SearchClient } = await import("@repo/vault-client/search");
  const client = new SearchClient(ctx.vaultPath);
  clientCache.set(ctx.vaultPath, client);
  return client;
}

/**
 * Generate a query embedding via OpenRouter embeddings API.
 * Returns null if no API key or on error (caller falls back to keyword).
 */
async function generateQueryEmbedding(
  query: string,
  apiKey: string,
): Promise<number[] | null> {
  try {
    const response = await fetch("https://openrouter.ai/api/v1/embeddings", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ model: EMBEDDING_MODEL, input: query }),
    });
    if (!response.ok) return null;
    const result = (await response.json()) as {
      data: Array<{ embedding: number[] }>;
    };
    return result.data[0]?.embedding ?? null;
  } catch {
    return null;
  }
}

// ── 1. vault_search (FTS5 Hybrid) ──────────────────────────────────

export const VaultSearchTool: HQTool<
  { query: string; mode?: "keyword" | "semantic" | "hybrid"; limit?: number },
  any
> = {
  name: "vault_search",
  description:
    "Search the local Obsidian vault using SQLite FTS5 (keyword) or semantic hybrid search. " +
    "Keyword mode is instant. Hybrid/semantic modes generate an embedding via OpenRouter and " +
    "combine it with FTS5 results (60% semantic + 40% keyword). Falls back to keyword if no API key.",
  tags: ["vault", "search", "query", "find", "knowledge"],
  schema: Type.Object({
    query: Type.String({ description: "Search query" }),
    mode: Type.Optional(
      Type.Union(
        [
          Type.Literal("keyword"),
          Type.Literal("semantic"),
          Type.Literal("hybrid"),
        ],
        {
          description:
            "Search mode. 'keyword' uses FTS5 only (fastest). 'hybrid' combines FTS5 + semantic (best quality). 'semantic' uses embeddings only. Default: keyword.",
        },
      ),
    ),
    limit: Type.Optional(
      Type.Number({ description: "Result limit (max 50). Default 10." }),
    ),
  }),
  async execute(input, ctx) {
    const client = await getSearchClient(ctx);
    const mode = input.mode || "keyword";
    const limit = Math.min(input.limit || 10, 50);

    if (mode === "keyword") {
      return client.keywordSearch(input.query, limit);
    }

    // For hybrid/semantic, try to generate a query embedding
    if (!ctx.openrouterApiKey) {
      // No API key — fall back to keyword silently
      return client.keywordSearch(input.query, limit);
    }

    const queryEmbedding = await generateQueryEmbedding(
      input.query,
      ctx.openrouterApiKey,
    );

    if (mode === "semantic") {
      if (!queryEmbedding) {
        return client.keywordSearch(input.query, limit);
      }
      return client.semanticSearch(queryEmbedding, limit);
    }

    // hybrid (default when API key available)
    return client.hybridSearch(input.query, queryEmbedding, limit);
  },
};

// ── 2. vault_read ───────────────────────────────────────────────────

export const VaultReadTool: HQTool<{ path: string; parseFrontmatter?: boolean }, any> = {
  name: "vault_read",
  description: "Read a note from the vault by vault-relative path.",
  tags: ["vault", "read", "file", "note"],
  schema: Type.Object({
    path: Type.String({ description: "Vault-relative path (e.g. Notebooks/Projects/ProjectA.md)" }),
    parseFrontmatter: Type.Optional(Type.Boolean({ description: "Whether to parse YAML frontmatter. Default true." }))
  }),
  async execute(input, ctx) {
    const fullPath = resolveVaultPath(ctx.vaultPath, input.path);
    if (!fs.existsSync(fullPath)) {
      throw new Error(`File not found in vault: ${input.path}`);
    }
    const content = fs.readFileSync(fullPath, "utf-8");
    if (input.parseFrontmatter !== false) {
      const { data, content: body } = matter(content);
      return { frontmatter: data, content: body };
    }
    return { content };
  }
};

// ── 3. vault_context (Structured System) ──────────────────────────

export const VaultContextTool: HQTool<{}, any> = {
  name: "vault_context",
  description: "Read core system files (SOUL, MEMORY, PREFERENCES, HEARTBEAT) as structured JSON.",
  tags: ["vault", "context", "system", "brain"],
  schema: Type.Object({}),
  async execute(_, ctx) {
    const systemDir = path.join(ctx.vaultPath, "_system");
    const files = ["SOUL.md", "MEMORY.md", "PREFERENCES.md", "HEARTBEAT.md"];
    const results: Record<string, any> = {};

    for (const f of files) {
      const fp = path.join(systemDir, f);
      if (fs.existsSync(fp)) {
        const raw = fs.readFileSync(fp, "utf-8");
        const { data, content } = matter(raw);
        results[f.replace(".md", "")] = { frontmatter: data, content };
      }
    }
    return results;
  }
};

// ── 4. vault_list ──────────────────────────────────────────────────

export const VaultListTool: HQTool<{ path?: string; recursive?: boolean }, any> = {
  name: "vault_list",
  description: "List files and directories in the vault.",
  tags: ["vault", "list", "directory", "ls"],
  schema: Type.Object({
    path: Type.Optional(Type.String({ description: "Vault-relative path. Default root." })),
    recursive: Type.Optional(Type.Boolean({ description: "List recursively. Default false." }))
  }),
  async execute(input, ctx) {
    const targetDir = resolveVaultPath(ctx.vaultPath, input.path || "");
    if (!fs.existsSync(targetDir)) return [];

    const listEntries = (dir: string, base: string): any[] => {
      const entries = fs.readdirSync(dir, { withFileTypes: true });
      let results: any[] = [];
      for (const entry of entries) {
        const relPath = path.join(base, entry.name);
        results.push({
          name: entry.name,
          path: relPath,
          isDir: entry.isDirectory(),
          size: entry.isFile() ? fs.statSync(path.join(dir, entry.name)).size : undefined
        });
        if (input.recursive && entry.isDirectory()) {
          results = results.concat(listEntries(path.join(dir, entry.name), relPath));
        }
      }
      return results;
    };

    return listEntries(targetDir, input.path || "");
  }
};

// ── 5. vault_batch_read ───────────────────────────────────────────

export const VaultBatchReadTool: HQTool<{ paths: string[] }, any> = {
  name: "vault_batch_read",
  description: "Read up to 20 files from the vault in a single call.",
  tags: ["vault", "batch", "read", "bulk"],
  schema: Type.Object({
    paths: Type.Array(Type.String(), { maxItems: 20, description: "List of vault-relative paths" })
  }),
  async execute(input, ctx) {
    const results: Record<string, any> = {};
    for (const p of input.paths) {
      const fp = resolveVaultPath(ctx.vaultPath, p);
      if (fs.existsSync(fp)) {
        const { data, content } = matter(fs.readFileSync(fp, "utf-8"));
        results[p] = { frontmatter: data, content };
      }
    }
    return results;
  }
};

// ── 6. vault_write_note ───────────────────────────────────────────

export const VaultWriteNoteTool: HQTool<{ path: string; content: string; frontmatter?: Record<string, any> }, any> = {
  name: "vault_write_note",
  description: "Create or overwrite a note in the vault. Only allowed in Notebooks/ or other designated areas.",
  tags: ["vault", "write", "note", "create"],
  requiresWriteAccess: true,
  schema: Type.Object({
    path: Type.String({ description: "Vault-relative path (must start with Notebooks/)" }),
    content: Type.String({ description: "Note body" }),
    frontmatter: Type.Optional(Type.Any({ description: "YAML frontmatter metadata" }))
  }),
  async execute(input, ctx) {
    if (!input.path.startsWith("Notebooks/")) {
      throw new Error("Writing outside Notebooks/ is restricted for security.");
    }
    const fullPath = resolveVaultPath(ctx.vaultPath, input.path);
    const dir = path.dirname(fullPath);
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });

    const fileContent = matter.stringify(input.content, input.frontmatter || {});
    fs.writeFileSync(fullPath, fileContent);
    return { success: true, path: input.path };
  }
};

// ── 7. vault_create_job ───────────────────────────────────────────

export const VaultCreateJobTool: HQTool<{ instruction: string; priority?: number; type?: string }, any> = {
  name: "vault_create_job",
  description: "Create a new job in the vault queue (_jobs/pending/).",
  tags: ["vault", "job", "create", "task", "orchestration"],
  requiresWriteAccess: true,
  schema: Type.Object({
    instruction: Type.String({ description: "Task instruction for the agent" }),
    priority: Type.Optional(Type.Number({ description: "Job priority (1-100). Default 50." })),
    type: Type.Optional(Type.String({ description: "Job type (background, rpc, interactive). Default background." }))
  }),
  async execute(input, ctx) {
    const jobId = `job-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
    const jobFilename = `${jobId}.md`;
    const fullPath = path.join(ctx.vaultPath, "_jobs", "pending", jobFilename);

    const frontmatter = {
      jobId,
      status: "pending",
      priority: input.priority || 50,
      type: input.type || "background",
      createdAt: new Date().toISOString()
    };

    const fileContent = matter.stringify(`# Instruction\n\n${input.instruction}`, frontmatter);
    const dir = path.dirname(fullPath);
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
    
    fs.writeFileSync(fullPath, fileContent);
    return { jobId, path: `_jobs/pending/${jobFilename}` };
  }
};

File: crates/hq-tools/src/vault.rs
//! Vault tools — 7 tools for reading, writing, searching, and managing vault notes.

use anyhow::{Result, bail};
use async_trait::async_trait;
use hq_db::Database;
use hq_vault::VaultClient;
use serde_json::{Value, json};
use std::sync::Arc;

use crate::registry::HqTool;

// ─── Helpers ────────────────────────────────────────────────────

/// Reject paths that escape the vault root via `..` or absolute prefixes.
fn validate_path(path: &str) -> Result<()> {
    if path.contains("..") || path.starts_with('/') || path.starts_with('\\') {
        bail!("path traversal not allowed: {}", path);
    }
    Ok(())
}

fn note_to_json(note: &hq_core::types::Note) -> Value {
    json!({
        "path": note.path,
        "title": note.title,
        "content": note.content,
        "tags": note.tags,
        "modified_at": note.modified_at.to_rfc3339(),
    })
}

// ─── VaultSearchTool ────────────────────────────────────────────

/// Keyword search via the FTS5 index in the database.
pub struct VaultSearchTool {
    db: Arc<Database>,
}

impl VaultSearchTool {
    pub fn new(db: Arc<Database>) -> Self {
        Self { db }
    }
}

#[async_trait]
impl HqTool for VaultSearchTool {
    fn name(&self) -> &str {
        "vault_search"
    }

    fn description(&self) -> &str {
        "Search vault notes by keyword query (FTS5). Returns matching paths, titles, and snippets."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "query": { "type": "string", "description": "Search query" },
                "limit": { "type": "integer", "description": "Max results (default 20)", "default": 20 }
            },
            "required": ["query"]
        })
    }

    fn category(&self) -> &str {
        "vault"
    }

    async fn execute(&self, args: Value) -> Result<Value> {
        let query = args
            .get("query")
            .and_then(|v| v.as_str())
            .unwrap_or_default()
            .to_string();
        let limit = args
            .get("limit")
            .and_then(|v| v.as_u64())
            .unwrap_or(20) as usize;

        let db = self.db.clone();
        let results: Vec<hq_core::types::SearchResult> = tokio::task::spawn_blocking(move || {
            db.with_conn(|conn| hq_db::search::keyword_search(conn, &query, limit))
        })
        .await??;

        let items: Vec<Value> = results
            .into_iter()
            .map(|r| {
                json!({ "path": r.note_path, "title": r.title, "snippet": r.snippet, "relevance": r.relevance })
            })
            .collect();

        Ok(json!({ "results": items, "count": items.len() }))
    }
}

// ─── VaultReadTool ──────────────────────────────────────────────

/// Read a single note by relative path.
pub struct VaultReadTool {
    vault: Arc<VaultClient>,
}

impl VaultReadTool {
    pub fn new(vault: Arc<VaultClient>) -> Self {
        Self { vault }
    }
}

#[async_trait]
impl HqTool for VaultReadTool {
    fn name(&self) -> &str {
        "vault_read"
    }

    fn description(&self) -> &str {
        "Read a vault note by its relative path. Returns title, content, tags, and metadata."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "path": { "type": "string", "description": "Relative path inside the vault (e.g. Notebooks/Projects/foo.md)" }
            },
            "required": ["path"]
        })
    }

    fn category(&self) -> &str {
        "vault"
    }

    async fn execute(&self, args: Value) -> Result<Value> {
        let path = args
            .get("path")
            .and_then(|v| v.as_str())
            .unwrap_or_default();
        validate_path(path)?;

        let vault = self.vault.clone();
        let path_owned = path.to_string();
        let note = tokio::task::spawn_blocking(move || vault.read_note(&path_owned)).await??;
        Ok(note_to_json(&note))
    }
}

// ─── VaultContextTool ───────────────────────────────────────────

/// Read the four system context files: SOUL, MEMORY, PREFERENCES, HEARTBEAT.
pub struct VaultContextTool {
    vault: Arc<VaultClient>,
}

impl VaultContextTool {
    pub fn new(vault: Arc<VaultClient>) -> Self {
        Self { vault }
    }
}

#[async_trait]
impl HqTool for VaultContextTool {
    fn name(&self) -> &str {
        "vault_context"
    }

    fn description(&self) -> &str {
        "Read system context files (SOUL.md, MEMORY.md, PREFERENCES.md, HEARTBEAT.md) from the vault root."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {},
            "required": []
        })
    }

    fn category(&self) -> &str {
        "vault"
    }

    async fn execute(&self, _args: Value) -> Result<Value> {
        let vault = self.vault.clone();
        let ctx = tokio::task::spawn_blocking(move || -> Result<Value> {
            let read = |name: &str| -> String {
                vault
                    .read_note(name)
                    .map(|n| n.content)
                    .unwrap_or_default()
            };
            Ok(json!({
                "soul": read("SOUL.md"),
                "memory": read("MEMORY.md"),
                "preferences": read("PREFERENCES.md"),
                "heartbeat": read("HEARTBEAT.md"),
            }))
        })
        .await??;
        Ok(ctx)
    }
}

// ─── VaultListTool ──────────────────────────────────────────────

/// List markdown files in a vault directory.
pub struct VaultListTool {
    vault: Arc<VaultClient>,
}

impl VaultListTool {
    pub fn new(vault: Arc<VaultClient>) -> Self {
        Self { vault }
    }
}

#[async_trait]
impl HqTool for VaultListTool {
    fn name(&self) -> &str {
        "vault_list"
    }

    fn description(&self) -> &str {
        "List markdown files in a vault directory. Optionally recurse into subdirectories."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "directory": { "type": "string", "description": "Relative directory to list (default: root)", "default": "" },
                "recursive": { "type": "boolean", "description": "Recurse into subdirectories", "default": false }
            },
            "required": []
        })
    }

    fn category(&self) -> &str {
        "vault"
    }

    async fn execute(&self, args: Value) -> Result<Value> {
        let dir = args
            .get("directory")
            .and_then(|v| v.as_str())
            .unwrap_or("");
        let recursive = args
            .get("recursive")
            .and_then(|v| v.as_bool())
            .unwrap_or(false);
        validate_path(dir)?;

        let vault = self.vault.clone();
        let dir_owned = dir.to_string();
        let paths = tokio::task::spawn_blocking(move || {
            if recursive {
                vault.list_notes_recursive(&dir_owned)
            } else {
                vault.list_notes(&dir_owned)
            }
        })
        .await??;

        Ok(json!({ "paths": paths, "count": paths.len() }))
    }
}

// ─── VaultBatchReadTool ─────────────────────────────────────────

/// Read up to 20 notes at once.
pub struct VaultBatchReadTool {
    vault: Arc<VaultClient>,
}

impl VaultBatchReadTool {
    pub fn new(vault: Arc<VaultClient>) -> Self {
        Self { vault }
    }
}

#[async_trait]
impl HqTool for VaultBatchReadTool {
    fn name(&self) -> &str {
        "vault_batch_read"
    }

    fn description(&self) -> &str {
        "Read up to 20 vault notes at once. Returns an array of notes (or errors for missing paths)."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "paths": {
                    "type": "array",
                    "items": { "type": "string" },
                    "description": "Array of relative paths (max 20)",
                    "maxItems": 20
                }
            },
            "required": ["paths"]
        })
    }

    fn category(&self) -> &str {
        "vault"
    }

    async fn execute(&self, args: Value) -> Result<Value> {
        let paths: Vec<String> = args
            .get("paths")
            .and_then(|v| serde_json::from_value(v.clone()).ok())
            .unwrap_or_default();

        if paths.len() > 20 {
            bail!("batch_read limited to 20 paths, got {}", paths.len());
        }

        for p in &paths {
            validate_path(p)?;
        }

        let vault = self.vault.clone();
        let results = tokio::task::spawn_blocking(move || -> Vec<Value> {
            paths
                .iter()
                .map(|p| match vault.read_note(p) {
                    Ok(note) => note_to_json(&note),
                    Err(e) => json!({ "path": p, "error": e.to_string() }),
                })
                .collect()
        })
        .await?;

        Ok(json!({ "notes": results, "count": results.len() }))
    }
}

// ─── VaultWriteNoteTool ─────────────────────────────────────────

/// Write a note, restricted to the `Notebooks/` prefix.
pub struct VaultWriteNoteTool {
    vault: Arc<VaultClient>,
}

impl VaultWriteNoteTool {
    pub fn new(vault: Arc<VaultClient>) -> Self {
        Self { vault }
    }
}

#[async_trait]
impl HqTool for VaultWriteNoteTool {
    fn name(&self) -> &str {
        "vault_write_note"
    }

    fn description(&self) -> &str {
        "Write or update a vault note. Path must start with Notebooks/ (safety restriction)."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "path": { "type": "string", "description": "Relative path starting with Notebooks/" },
                "title": { "type": "string", "description": "Note title" },
                "content": { "type": "string", "description": "Markdown content body" },
                "tags": { "type": "array", "items": { "type": "string" }, "description": "Optional tags", "default": [] }
            },
            "required": ["path", "title", "content"]
        })
    }

    fn category(&self) -> &str {
        "vault"
    }

    async fn execute(&self, args: Value) -> Result<Value> {
        let path = args
            .get("path")
            .and_then(|v| v.as_str())
            .unwrap_or_default()
            .to_string();
        let title = args
            .get("title")
            .and_then(|v| v.as_str())
            .unwrap_or_default()
            .to_string();
        let content = args
            .get("content")
            .and_then(|v| v.as_str())
            .unwrap_or_default()
            .to_string();
        let tags: Vec<String> = args
            .get("tags")
            .and_then(|v| serde_json::from_value(v.clone()).ok())
            .unwrap_or_default();

        validate_path(&path)?;
        if !path.starts_with("Notebooks/") {
            bail!("vault_write_note is restricted to Notebooks/ prefix, got: {path}");
        }

        let mut frontmatter = std::collections::HashMap::new();
        frontmatter.insert(
            "title".to_string(),
            serde_yaml::Value::String(title.clone()),
        );
        if !tags.is_empty() {
            frontmatter.insert(
                "tags".to_string(),
                serde_yaml::Value::Sequence(
                    tags.iter()
                        .map(|t| serde_yaml::Value::String(t.clone()))
                        .collect(),
                ),
            );
        }

        let note = hq_core::types::Note {
            title: title.clone(),
            content,
            path: path.clone(),
            frontmatter,
            note_type: None,
            tags,
            pinned: false,
            source: None,
            embedding_status: None,
            created_at: None,
            updated_at: None,
            modified_at: chrono::Utc::now(),
        };

        let vault = self.vault.clone();
        let path_ret = path.clone();
        tokio::task::spawn_blocking(move || vault.write_note(&path, &note)).await??;

        Ok(json!({ "ok": true, "path": path_ret }))
    }
}

// ─── VaultCreateJobTool ─────────────────────────────────────────

/// Create a new job in the vault pending queue.
pub struct VaultCreateJobTool {
    vault: Arc<VaultClient>,
}

impl VaultCreateJobTool {
    pub fn new(vault: Arc<VaultClient>) -> Self {
        Self { vault }
    }
}

#[async_trait]
impl HqTool for VaultCreateJobTool {
    fn name(&self) -> &str {
        "vault_create_job"
    }

    fn description(&self) -> &str {
        "Create a new background job in the vault pending queue."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "instruction": { "type": "string", "description": "Job instruction / prompt" },
                "model": { "type": "string", "description": "Optional model override" }
            },
            "required": ["instruction"]
        })
    }

    fn category(&self) -> &str {
        "vault"
    }

    async fn execute(&self, args: Value) -> Result<Value> {
        let instruction = args
            .get("instruction")
            .and_then(|v| v.as_str())
            .unwrap_or_default()
            .to_string();
        let model = args
            .get("model")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string());

        let vault = self.vault.clone();
        let job = tokio::task::spawn_blocking(move || {
            vault.create_job(&instruction, model.as_deref())
        })
        .await??;

        Ok(json!({
            "ok": true,
            "job_id": job.id,
            "status": "pending",
        }))
    }
}
